src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp

```diff
@@ -71,7 +71,7 @@ address ProgrammableUpcallHandler::generate_upcall_stub(jobject rec, jobject jab
 
   // Capture prev stack pointer (stack arguments base)
   __ add(rscratch1, rfp, 16);   // Skip saved FP and LR
-  __ str(rscratch1, Address(sp, layout.stack_args));
+  __ str(rscratch1, Address(sp, checked_cast<int>(layout.stack_args)));  // x86 casts to int also
```

Comment:
```suggestion
  __ Address slot = __ legitimize_address(Address(sp, checked_cast<int>(layout.stack_args)), wordSize, rscratch2);
  __ str(rscratch1, slot);  // x86 casts to int also
```
I think this is a real bug: the range of a stack arg from SP can exceed that of the maximum offset of a STR instruction!
Wherever there's a dubious cast there's probably a bug...

Reply:
All of this may seem tedious and pedantic, but we have had failures in production caused by stack pointer offsets exceeding the 12-bit range of a STR instruction.

Reply:
Yes, I agree and didn't really know how to fix it so it would compile (if it actually compiles now) and not be UB.  I'm close to giving up!  Thanks for the code change.

Reply:
It does seem pedantic which makes it really difficult, but it's important to get this right.  These kinds of bugs are horrible to debug.

Reply:
OK, so if we have the `Address(reg_offset)` overloads for `int`, `int64_t`, and uint64_t` we're good, and by calling the right overload we get rid of the need to cast the `layout.stack_args` offset to `int` here.

Reply:
This is what I've got:
```

  Address(Register r)
    : _base(r), _index(noreg), _offset(0), _mode(base_plus_offset), _target(0) { }
  Address(Register r, int o)
    : _base(r), _index(noreg), _offset(o), _mode(base_plus_offset), _target(0) { }
  Address(Register r, int64_t o)
    : _base(r), _index(noreg), _offset(o), _mode(base_plus_offset), _target(0) { }
  Address(Register r, uint64_t o)
    : _base(r), _index(noreg), _offset(o), _mode(base_plus_offset), _target(0) { }
  Address(Register r, ByteSize disp)
    : Address(r, in_bytes(disp)) { }

```

Reply:
I added ptrdiff_t but that's the same as int64_t.  I'll fix it.

Reply:
I need ptrdiff_t additionally to get this to compile on our macosx-aarch64 platform.